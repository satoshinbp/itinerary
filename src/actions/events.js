import db from '../firebase/firebase'

// ADD_EVENT
export const addEvent = event => ({
  type: 'ADD_EVENT',
  event
})

export const startAddEvent = (tripId, eventData) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    const eventsRef = db.collection('users').doc(uid).collection('trips').doc(tripId).collection('events')
    const { title, date, startTime, endTime, location, note } = eventData
    const event = { title, date, startTime, endTime, location, note, user: uid, tripId }

    return eventsRef.add(event).then(snapshot => (
      dispatch(addEvent({
        id: snapshot.id,
        ...event
      }))
    ))
  }
}

// EDIT_EVENT
export const editEvent = (id, updates) => ({
  type: 'EDIT_EVENT',
  id,
  updates
})

export const startEditEvent = (tripId, id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    const eventsRef = db.collection('users').doc(uid).collection('trips').doc(tripId).collection('events').doc(id)
    return eventsRef.update(updates).then(() => dispatch(editEvent(id, updates)))
  }
}

// REMOVE_EVENT
export const removeEvent = id => ({
  type: 'REMOVE_EVENT',
  id
})

export const startRemoveEvent = (tripId, id) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    const eventsRef = db.collection('users').doc(uid).collection('trips').doc(tripId).collection('events').doc(id)
    return eventsRef.delete().then(() => dispatch(removeEvent(id)))
  }
}

// SET_EVENTS
export const setEvents = events => ({
  type: 'SET_EVENTS',
  events
})

export const startSetEvents = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    const events = []

    return db.collectionGroup('events').where("user", "==", uid).get().then(querySnapshot => {
      querySnapshot.forEach(snapshot => {
        events.push({
          id: snapshot.id,
          tripId: snapshot.ref.parent.parent.id,
          user: uid,
          ...snapshot.data()
        })
      })
      dispatch(setEvents(events))
    })
  }
}