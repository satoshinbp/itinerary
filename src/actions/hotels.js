export const addHotel = (
  {
    id='',
    tripId = '',
    name = '',
    checkInDate = 0,
    checkOutDate = 0,
    checkInTime = 0,
    checkOutTime = 0,
    ETA = 0,
    ETD = 0,
    location ='',
    note = ''
  } = {}
) => ({
  type: 'ADD_HOTEL',
  hotel: {
    id,
    tripId,
    name,
    checkInDate,
    checkOutDate,
    checkInTime,
    checkOutTime,
    ETA,
    ETD,
    location,
    note
  }
})

export const editHotel = (id, updates) => ({
  type: 'EDIT_HOTEL',
  id,
  updates
})

export const removeHotel = (id = '') => ({
  type: 'REMOVE_HOTEL',
  id
})