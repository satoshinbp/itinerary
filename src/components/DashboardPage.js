import React from 'react';
import { Link } from 'react-router-dom';
import TripList from './TripList';
import TripListFilters from './TripListFilters';

const DashboardPage = () => (
  <div className="container">
    <div className="page-header">
      <Link className="btn-push" to="/create">Add Trip</Link>
      <TripListFilters />
    </div>
    <TripList />
  </div>
);

export default DashboardPage;