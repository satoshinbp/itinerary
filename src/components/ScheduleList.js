import React from 'react'
import { connect } from 'react-redux'
import { v4 as uuidv4 } from 'uuid'
import moment from 'moment'
import EventItem from './EventItem'
import HotelItem from './HotelItem'

const ScheduleList = (props) => {
  return (
    <React.Fragment>
      <div className="schedule-list-header">
        <div>{props.date.format('YYYY/MM/DD')}</div>
        <div>
          <button
            className="btn-push btn-push--small"
            onClick={() => {
              props.setEventId(uuidv4())
              props.setEventDate(props.date)
            }}
          >
            Add Event
          </button>
          <button
            className="btn-push btn-push--small"
            onClick={() => {
              props.setHotelId(uuidv4())
              props.setEventDate(props.date)
            }}
          >
            Add Hotel
        </button>
        </div>
      </div>
      <div className="schedule-list-body">
        {
          props.hotelsOut.length !== 0 && props.hotelsOut.map(hotel =>
            <HotelItem key={hotel.id} {...hotel} setHotelId={props.setHotelId} date={props.date} inOrOut="out" />)
        }
        {
          props.events.length !== 0 && props.events.map(event =>
            <EventItem key={event.id} {...event} setEventId={props.setEventId} />)
        }
        {
          props.hotelsIn.length !== 0 && props.hotelsIn.map(hotel =>
            <HotelItem key={hotel.id} {...hotel} setHotelId={props.setHotelId} date={props.date} inOrOut="in"/>)
        }
      </div>
    </React.Fragment>
  )
}

const mapStateToProps = (state, props) => ({
  events: state.events
    .filter(event =>
      event.tripId === props.tripId && moment(event.date).isSame(props.date, 'day'))
    .sort((a, b) =>
      moment(a.startTime).isAfter(moment(b.startTime)) ? 1 : -1),
  hotelsIn: state.hotels
    .filter(hotel =>
      hotel.tripId === props.tripId && props.date.isBetween(moment(hotel.checkInDate), moment(hotel.checkOutDate), 'day', '[)')),
  hotelsOut: state.hotels
    .filter(hotel =>
      hotel.tripId === props.tripId && props.date.isBetween(moment(hotel.checkInDate), moment(hotel.checkOutDate), 'day', '(]'))
})

export default connect(mapStateToProps)(ScheduleList)