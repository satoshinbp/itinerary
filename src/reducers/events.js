const eventsReducerDefaultState = []

export default (state = eventsReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EVENT':
      return [
        ...state,
        action.event
      ]
    case 'EDIT_EVENT':
      return state.map((event) => {
        if (event.id === action.id) {
          return {
            ...event,
            ...action.updates
          }
        } else {
          return event
        }
      })
    case 'REMOVE_EVENT':
      return state.filter(({ id }) => id !== action.id)
    default:
      return state
  }
}