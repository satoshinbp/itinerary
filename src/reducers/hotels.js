const hotelsReducerDefaultState = []

export default (state = hotelsReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_HOTEL':
      return [
        ...state,
        action.hotel
      ]
    case 'EDIT_HOTEL':
      return state.map((hotel) => {
        if (hotel.id === action.id) {
          return {
            ...hotel,
            ...action.updates
          }
        } else {
          return hotel
        }
      })
    case 'REMOVE_HOTEL':
      return state.filter(({ id }) => id !== action.id)
    default:
      return state
  }
}