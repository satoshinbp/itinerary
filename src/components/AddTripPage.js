import React from 'react';
import { connect } from 'react-redux';
import TripForm from './TripForm';
import { addTrip } from '../actions/trips';

const AddTripPage = (props) => (
  <div className="container">
    <TripForm
      onSubmit={(trip) => {
        props.dispatch(addTrip(trip));
        props.history.push('/');
      }} />
  </div>
);

export default connect()(AddTripPage);