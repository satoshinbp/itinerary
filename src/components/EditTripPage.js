import React from 'react';
import { connect } from 'react-redux';
import TripForm from './TripForm';
import { editTrip, removeTrip } from '../actions/trips';

const EditTripPage = (props) => {
  return (
    <div className="container">
      <TripForm
        trip={props.trip}
        onSubmit={(trip) => {
          props.dispatch(editTrip(props.trip.id, trip));
          props.history.push('/');
        }} />
      <button
        className="btn-push btn-push--secondary"
        onClick={() => {
          props.dispatch(removeTrip({ id: props.trip.id }));
          props.history.push('/');
        }}>Remove Expense</button>
    </div>
  );
};

const mapStateToProps = (state, props) => ({
  trip: state.trips.find((trip) => trip.id === props.match.params.id)
});

export default connect(mapStateToProps)(EditTripPage);
