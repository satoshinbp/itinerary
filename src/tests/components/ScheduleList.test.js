import React from 'react'
import { shallow } from 'enzyme'
import moment from 'moment'
import { ScheduleList } from '../../components/ScheduleList'
import events from '../fixtures/events'
import hotels from '../fixtures/hotels'

let setEventId, setHotelId, setDateOnEdit, wrapper
const date = moment().add(1, 'day')

describe('for 2nd day of trips[0]', () => {
  beforeEach(() => {
    setEventId = jest.fn()
    setHotelId = jest.fn()
    setDateOnEdit = jest.fn()
    wrapper = shallow(
      <ScheduleList
        date={date}
        events={[events[0]]}
        hotelsIn={[hotels[0]]}
        hotelsOut={[hotels[2]]}
        setEventId={setEventId}
        setHotelId={setHotelId}
        setDateOnEdit={setDateOnEdit}
      />
    )
  })

  test('should render ScheduleList correctly', () => {
    expect(wrapper).toMatchSnapshot()
  })

  test('should set event id and date on add event click', () => {
    wrapper.find('button').at(0).simulate('click')
    expect(setEventId).toBeCalledWith(true)
    expect(setDateOnEdit).toBeCalledWith(date)
  })

  test('should set hotel id and date on add hotel click', () => {
    wrapper.find('button').at(0).simulate('click')
    expect(setEventId).toBeCalledWith(true)
    expect(setDateOnEdit).toBeCalledWith(date)
  })
})

test('for 2nd day of trips[0] without events, hotels', () => {
  setEventId = jest.fn()
  setHotelId = jest.fn()
  setDateOnEdit = jest.fn()
  wrapper = shallow(
    <ScheduleList
      date={date}
      events={[]}
      hotelsIn={[]}
      hotelsOut={[]}
      setEventId={setEventId}
      setHotelId={setHotelId}
      setDateOnEdit={setDateOnEdit}
    />
  )
  expect(wrapper).toMatchSnapshot()
})