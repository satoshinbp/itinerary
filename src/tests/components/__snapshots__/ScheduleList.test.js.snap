// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`for 2nd day of trips[0] should render ScheduleList correctly 1`] = `
<Fragment>
  <div
    className="schedule-list-header"
  >
    <div>
      1970/01/02
    </div>
    <div>
      <button
        className="btn-push btn-push--small"
        onClick={[Function]}
      >
        Add Event
      </button>
      <button
        className="btn-push btn-push--small"
        onClick={[Function]}
      >
        Add Hotel
      </button>
    </div>
  </div>
  <div
    className="schedule-list-body"
  >
    <Connect(HotelItem)
      ETA={18000000}
      ETD={3600000}
      checkInDate={0}
      checkInTime={18000000}
      checkOutDate={86400000}
      checkOutTime={3600000}
      date={"1970-01-02T00:00:00.000Z"}
      id="3"
      inOrOut="out"
      key="3"
      location=""
      name="Hawai Travellars Inn"
      note="Early Check-in"
      setHotelId={[MockFunction]}
      tripId="a"
    />
    <Connect(EventItem)
      date={86400000}
      endTime={25200000}
      id="1"
      key="1"
      location="Waikiki"
      note=""
      setEventId={[MockFunction]}
      startTime={14400000}
      title="Shopping"
      tripId="a"
    />
    <Connect(HotelItem)
      ETA={32400000}
      ETD={0}
      checkInDate={86400000}
      checkInTime={21600000}
      checkOutDate={172800000}
      checkOutTime={7200000}
      date={"1970-01-02T00:00:00.000Z"}
      id="1"
      inOrOut="in"
      key="1"
      location="Waikiki"
      name="Waikiki Resort Hotel"
      note=""
      setHotelId={[MockFunction]}
      tripId="a"
    />
  </div>
</Fragment>
`;

exports[`for 2nd day of trips[0] without events, hotels 1`] = `
<Fragment>
  <div
    className="schedule-list-header"
  >
    <div>
      1970/01/02
    </div>
    <div>
      <button
        className="btn-push btn-push--small"
        onClick={[Function]}
      >
        Add Event
      </button>
      <button
        className="btn-push btn-push--small"
        onClick={[Function]}
      >
        Add Hotel
      </button>
    </div>
  </div>
  <div
    className="schedule-list-body"
  />
</Fragment>
`;
